### PR TITLE
Create separate generic for schema definition

### PR DESCRIPTION
**Pull request by vkarpov15**
**Summary**

Re: #13523, #13772.

The current implementation of automatic type inference makes it tricky for us to automatically infer the correct hydrated document type. Because automatic type inference infers the `DocType`, which is a mix of the hydrated document type and the raw doc type: hydrated for arrays and maps, raw for everything else, including nested paths. And then we infer the schema definition type from the `DocType`, which means we lose whether object paths are [nested paths or subdocument paths](https://mongoosejs.com/docs/subdocs.html#subdocuments-versus-nested-paths).

This PR presents one approach that can help us move towards automatically inferring the raw doc type and the hydrated doc type correctly: getting the schema definition and inferring the raw doc type from that. In theory, if we have the schema definition type, we should be able to infer the hydrated document type from that.

The downside is this approach seems to require adding `as const` to schema definitions if you want to use `type: 'String'` or `type: 'Number'` as opposed to `type: String` or `type: Number`. Also has some issues with context for default functions, but those issues are present to some extent even without this PR and have easy workarounds.

cc @mohammad0-0ahmad 

**Examples**

### CODE REVIEW

File: test/types/schema.test.ts [hunk 1]
```diff
@@ -1007,7 +1014,7 @@ function gh12869() {
 
   const dbExample = new Schema(
     {
-      active: { type: String, enum: ['foo', 'bar'], required: true }
+      active: { type: String, enum: ['foo', 'bar'] as ['foo', 'bar'], required: true }
```
Comment (hasezoey) [inline test/types/schema.test.ts:1017]: is this casting really required instead of just `as const`?

File: test/types/virtuals.test.ts [hunk 2]
```diff
@@ -95,14 +95,15 @@ async function autoTypedVirtuals() {
   const testSchema = new Schema({
     email: {
       type: String,
-      required: [true, 'email is required']
+      required: [true, 'email is required'] as [true, string]
```
Comment (hasezoey) [inline test/types/virtuals.test.ts:98]: are those array castings really required instead of `as const`?

File: test/types/virtuals.test.ts [hunk 3]
```diff
@@ -95,14 +95,15 @@ async function autoTypedVirtuals() {
...
     }
   }, {
     virtuals: {
       domain: {
         get() {
           expectType<Document<any, any, { email: string }> & AutoTypedSchemaType>(this);
-          return this.email.slice(this.email.indexOf('@') + 1);
+          const email = this.email;
+          return email.slice(email.indexOf('@') + 1);
```
Comment (hasezoey) [inline test/types/virtuals.test.ts:106]: why did this change? to inspect the type?

**Review by hasezoey — COMMENTED** [review]
Looks OK, it is really not ideal to require explicit casting or redundant `as const` in some places, as long as it is limited to the string representation (of primitives) it should be OK

i have not yet tested this change (or 8.0 in general) against typegoose for type issues

File: test/types/virtuals.test.ts (hunk 2 above)
Comment (vkarpov15) [reply test/types/virtuals.test.ts:98]: Unfortunately yes. I think that's more of a TypeScript 5 issue though, similar to #13514.

File: test/types/schema.test.ts (hunk 1 above)
Comment (vkarpov15) [reply test/types/schema.test.ts:1017]: This particular one isn't required, I'm changing.

**Pull request closed without merging**